corehq/apps/users/models.py

```diff
@@ -688,14 +697,24 @@ def has_permission(self, domain, permission, data=None, restrict_global_admin=Fa
             elif self.is_domain_admin(domain):
                 return True
 
-        dm = self.get_domain_membership(domain)
-        if dm:
+        def _domain_membership_has_permission(membership):
+            if not membership:
+                return False
             # an admin has access to all features by default, restrict that if needed
-            if dm.is_admin and restrict_global_admin:
+            if membership.is_admin and restrict_global_admin:
                 return False
-            return dm.has_permission(permission, data)
-        else:
-            return False
+            return membership.has_permission(permission, data)
+
+        if _domain_membership_has_permission(self.get_domain_membership(domain)):
+            return True
+
+        master_link = get_domain_master_link(domain)
+        if master_link and not master_link.is_remote:
+            if toggles.ENTERPRISE_LINKED_DOMAINS.enabled(master_link.master_domain):
+                if _domain_membership_has_permission(self.get_domain_membership(master_link.master_domain)):
```

Comment:
This means that a user will have all permissions they have in the master domain in all the linked ones, regardless of role in the master (ie not just admin)?

Is it a true statement to say that you can add permissions in a linked domain (by giving them an explicit role with new permissions), but not remove them, since it will fall back to checking the master if they do not have the permission in the linked domain?

That might be slightly confusing, if a user is given an intentionally restricted role in a given linked domain, but inherits a bunch of permissions from the master domain. It is at least worth documenting that somewhere, if that is true. 

If it solves the main use case, it might also be worth returning the value from the local domain membership if it exists, rather than just checking if the membership grants a permission, and moving on if not.

Reply:
What you're describing is how it works, yes. Let's offline with Dev about if that's how it should work.

Reply:
Updated in c5a96b38351